### PR TITLE
Add bulk risk acceptance API

### LINKED ISSUE #1889: Add REST API for risk acceptance
**Is your feature request related to a problem? Please describe**
I have a generated greylist of acceptable findings when located by specific tools on specific products. This seems like a natural use case for accepted findings, though the only REST APIs available for manipulating risk acceptance is if you already have a risk acceptance id.

**Describe the solution you'd like**
I'd like REST APIs to bulk-import risk acceptance for:
* A specific test
* A specific engagement and optionally which test types to apply them to
* A specific product and optionally which test types and version(s) to apply them to

**Describe alternatives you've considered**
External script to import greylist and send out finding REST API calls to mark them as mitigated instead. I'd prefer to extend the risk acceptance feature to be more useful if possible, though.

**Additional context**
Example greylist (has entries that apply to a few different test types, some of which already have parsers, and others I'll be making parsers for separately): https://dccscr.dsop.io/dsop/dccscr-whitelists/blob/master/redhat/ubi/ubi8/8.1/ubi8-8.1.greylist

### PR DESCRIPTION
**Pull request by jvz**
This adds two REST APIs: accept_risks on an engagement or a test. This
allows a list of CVEs (or other vulnerability identifiers) along with
justification/compensating control information and the approver of that
risk acceptance. These APIs can be used to bulk-import whitelists and
other risk acceptance criteria from external sources.

Closes #1889.

**Note: DefectDojo is now on Python3 and Django 2.2.1 Please submit your pull requests to the 'dev' branch as the 'legacy-python2.7' branch is only for bug fixes. Any new features submitted to the legacy branch will be ignored and closed.**

When submitting a pull request, please make sure you have completed the following checklist:

- [x] Your code is flake8 compliant 
- [x] Your code is python 3.5 compliant
- [ ] If this is a new feature and not a bug fix, you've included the proper documentation in the ReadTheDocs documentation folder. https://github.com/DefectDojo/Documentation/tree/master/docs or provide feature documentation in the PR.
- [x] Model changes must include the necessary migrations in the dojo/dd_migrations folder.
- [x] Add applicable tests to the unit tests.

### CODE REVIEW

**Comment by valentijnscholten** [comment]
Looks interesting, it's a usecase also on our list. "What if we want to risk accept a vulnerable dependency across all products?". So for me it would be even better if we extended this to product level and/or product_type level. Did you consider that? Or better yet support "instance level", i.e. accept all findings matching the cve across all products.

**Comment by jvz** [comment]
The original issue I filed mentioned applying to an entire product, though for consistency, it seems like it'd be simple enough to apply to product types as well.

Some of this is limited by the existing risk acceptance model as it's separate from finding, but then again, finding is quickly becoming a mega-model as it is.

My particular use case is to bulk-import externally-approved CVE whitelists like the one given in #1889, and without manually marking up the greylist data with specific mitigation types, it seemed like risk acceptance was the most appropriate one to use as it would only apply to findings that haven't already been deduplicated by previous engagements on the same product. The UI for risk acceptance is a bit lacking, though, but the searches for accepted findings work well enough that I can work around that UI (also, both the Jira and Sonar integration use risk acceptance similarly).

**Comment by valentijnscholten** [comment]
The PR only seems to support engagement and test level, not product?

To be honest I don't like the risk acceptance feature as it is. I understand it can be useful for formal risk acceptance, but more often I just like to triage my findings and accept them with a simple click/checkbox. So just a new 'Accepted' checkbox for a finding would be fine. But that's a whole other discussion.

**Comment by jvz** [comment]
As someone who came into security from an engineering perspective rather than infosec, I, too, find the risk acceptance UI to be overly formal. In my production usage, instead of risk acceptance for actual risk acceptance, we're marking a finding as mitigated without any status (i.e., uncheck all the boxes) and leaving a note that explains the justification/compensating control/whatever. This matches our workflow for when we mark a finding as a false positive, manually linking duplicates, or out of scope (i.e., all the statuses other than "Accepted").

We can definitely tease out an improved UX there in a separate PR, though! :)

**Comment by jvz** [comment]
Oh, and I didn't add it to product yet. I'll add it to product and product type.

**Comment by jvz** [comment]
Updated to allow for bulk acceptance on a product, product type, or just findings in general.

**Comment by madchap** [comment]
Hey @jvz 

You got some unresolved flake8 issues, please check the Travis build.

Here is a snippet

```
Found violations: dojo/api_v2/views.py
	F401 @ 1:1 - 'typing.List' imported but unused
	F401 @ 1:1 - 'typing.Optional' imported but unused
The command "./travis/script.sh" exited with 1.```

**Comment by madchap** [comment]
Re-opening, travis crapped out. All should be fine though.

**Comment by valentijnscholten** [comment]
I extended the unittests to confirm if my understanding was correct (it is). 

1) RiskAcceptances are linked in the database to *all* engagements, including those who do no have any findings that are accepted by these RA's. RA's should only be connected to engagements they are affecting.

2) RiskAcceptances are created "cross-product". This puts findings from product1 and product2 in the same RA risking being shown to unauthorized users. For example in the `view_engagement` view there is code that just retrieves all findings from RA's in that engagement and shows them all to the user. Same goes for engagements. People may see findings from version 1.0 when viewing engagement for version 2.0.

3) Performance
If the intent is to accept thouand(s) of findings by creating 1 risk acceptance for 1 CVE, combined with the above, this will lead to performance issues.
For example if there is a product with 1 engagement 1 unaccepted finding and 1 accepted finding. DD will then retrieve the risk acceptance that contains thousand(s) of findings just to show/detect 1 accepted finding. 

Suggestions:
1) Create risk acceptance objects per engagement.
2) Consider creating 1 risk acceptance object that accepts multiple CVEs, i.e. a whole list.

**Comment by madchap** [comment]
In the context of an automated run, I am not sure that linking having this broadly connected at the instance level makes any sense. I think we should remove this.

**Comment by jvz** [comment]
I removed it from the product type API. As for linking per engagement, I'm not exactly sure on the best way to refactor this right now as I've been away from the code for a while at this point.

**Comment by jvz** [comment]
And yes, you're right about the desire to have a per-CVE justification for things.

**Comment by madchap** [comment]
@jvz please fix the travis issues. Saw one, could be others after.

```
FAIL: test_finding_accept_risks (dojo.unittests.test_risk_acceptance_api.TestRiskAcceptanceApi)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/dojo/unittests/test_risk_acceptance_api.py", line 91, in test_finding_accept_risks
    self.assertEquals(self.engagement_2a.risk_acceptance.count(), 0)
AssertionError: 80 != 0
```

**Comment by madchap** [comment]
Also noticing there is no `DELETE` endpoints for `/accept_risks` on the various entities :thinking: 
How hard would it be to add that? Cheers.

**Comment by jvz** [comment]
Not sure why the test is failing. This was modified by @valentijnscholten, and I'm not entirely sure what's supposed to happen here.

**Comment by valentijnscholten** [comment]
I don't think I can make it more clear than my previous comment: https://github.com/DefectDojo/django-DefectDojo/pull/1904#issuecomment-595436150 Could you let me know which part is unclear?
It could be debated that seggregation on the product level might be sufficient, but in the current state I really think this should not be merged.

**Comment by jvz** [comment]
Oh I think I see where I got confused. You updated the unit test to show what you'd expect the behavior to be. Let me look into this.

**Comment by madchap** [comment]
My last manual test confirm RA are now scoped only to engagements.

**Comment by valentijnscholten** [comment]
Thanks, that was the idea yes :-)

I think the risk acceptance db object creation is now more what I was hoping for, but:

- The api endpoint for Product and ProdctType can still remain (be added back in), as long as the RA are created per engagement. I think it might be usefule to be able to use whitelist for certain products and not for others.
I think there is some remaining code anyway, i.e. risk_application_model_class for product etc.

- I still see the code iterating over ALL engagements, which I still don't understand. Once you have the findings you want to mitigate, I think you also have the engagements?
Something like for eng in [finding.test.engagement in findings_matching_cve]

- To find findings that are not yet accepted you can do something like
Finding.objects.annotate(ra_count=Count('risk_acceptance_set')).filter(c__eq=0)

**Comment by madchap** [comment]
@jvz, some of these suggestions you could take a look at to close this off?

Cheers :)

**Comment by jvz** [comment]
> Thanks, that was the idea yes :-)
> 
> I think the risk acceptance db object creation is now more what I was hoping for, but:
> 
>     * The api endpoint for Product and ProdctType can still remain (be added back in), as long as the RA are created per engagement. I think it might be usefule to be able to use whitelist for certain products and not for others.

I found that refactoring the code to support this would take too much time, so I just removed the endpoints. I'm not even using those endpoints in my use case (I would be using the engagement endpoint mostly), so I can't really justify additional work on that right now.

>       I think there is some remaining code anyway, i.e. risk_application_model_class for product etc.
> 
>     * I still see the code iterating over ALL engagements, which I still don't understand. Once you have the findings you want to mitigate, I think you also have the engagements?
>       Something like for eng in [finding.test.engagement in findings_matching_cve]

I'll look into this. That makes more sense, yes. :)

>     * To find findings that are not yet accepted you can do something like
>       Finding.objects.annotate(ra_count=Count('risk_acceptance_set')).filter(c__eq=0)

Neat! Will update PR later today.

**Comment by jvz** [comment]
> I still see the code iterating over ALL engagements, which I still don't understand. Once you have the findings you want to mitigate, I think you also have the engagements?
> Something like for eng in [finding.test.engagement in findings_matching_cve]

To avoid creating risk acceptance sets that link findings between different engagements or products or anything else, I could rearrange these queries to prevent that, though it starts to get messy. I could also update this to only go over active engagements or similar.

> To find findings that are not yet accepted you can do something like
>       Finding.objects.annotate(ra_count=Count('risk_acceptance_set')).filter(c__eq=0)

How is this any different from `Finding.unaccepted_open_findings` as is?

**Comment by madchap** [comment]
Hi @valentijnscholten,

I am trying to make these Travis tests pass, hopefully your last PR #2101 makes a difference :+1: 

Is there anything that you believe would be blocking here to merging this, that could not be done later on eventually? As I have some interest in this, I've been using it on a sandbox environment and it looks ok to me in its current state even though there could be future improvements.

Thanks.

**Comment by madchap** [comment]
@marianG would you care to provide your input on this one?

**Comment by madchap** [comment]
@jvz let's plan a future iteration after further dogfooding. Thank you.

File: dojo/risk_acceptance/api.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        for eng in models.Engagement.objects.filter(product=model):
+            add_risks_to_engagement(eng)
+    elif isinstance(model, models.Engagement):
+        add_risks_to_engagement(model)
+    elif isinstance(model, models.Test):
+        add_risks_to_engagement(model.engagement)
+    else:
+        for eng in models.Engagement.objects.all():
+            add_risks_to_engagement(eng)
+
```
[the remark this answers is not in the document]
Comment (jvz) [reply dojo/risk_acceptance/api.py]: I'm pretty sure I refactored the code to prevent cross-engagement contamination.

**Pull request merged**